[WebGL] Abstract attaching/detaching external images

djg opened the pull request:
#### 891ad78c430b6b02f4b5236174b500ba6f3523b3
<pre>
[WebGL] Abstract attaching/detaching external images
<a href="https://bugs.webkit.org/show_bug.cgi?id=256945">https://bugs.webkit.org/show_bug.cgi?id=256945</a>
rdar://109495127

Reviewed by Kimmo Kinnunen.

When the GPU process is enabled, the Web process no long has access to the APIs
required to create images passed from the UI process. This change added APIs to
allow attaching and detaching external images that can be represented by a
MachSendRight (either IOSurface or MTLSharedTexture), that can be used in the
Web or GPU process.

Support work for WebXR in GPUP.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
* Source/WebCore/PAL/pal/spi/cocoa/MetalSPI.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createAndBindExternalImage):
(WebCore::GraphicsContextGLANGLE::destroyEGLImage):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newSharedEventWithMachPort):
(WebCore::newSharedEvent):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::createAndBindSharedTextureToExternalImage):
(WebCore::GraphicsContextGLCocoa::createAndBindExternalImage):
(WebCore::GraphicsContextGLCocoa::attachIOSurfaceToSharedTexture): Deleted.
(WebCore::GraphicsContextGLCocoa::detachIOSurfaceFromSharedTexture): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::createAndBindExternalImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(destroyEGLImage):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder&lt;WebCore::GraphicsContextGL::ExternalImageSourceIOSurfaceHandle&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::GraphicsContextGL::ExternalImageSourceIOSurfaceHandle&gt;::decode):
(IPC::ArgumentCoder&lt;WebCore::GraphicsContextGL::ExternalImageSourceMTLSharedTextureHandle&gt;::encode):
(IPC::ArgumentCoder&lt;WebCore::GraphicsContextGL::ExternalImageSourceMTLSharedTextureHandle&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createAndBindExternalImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLImage):
* Tools/Scripts/generate-gpup-webgl:

Canonical link: <a href="https://commits.webkit.org/264712@main">https://commits.webkit.org/264712@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f554ae97b44dce1979058fdb0f0a12be8833c509

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/8405 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/8698 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/8915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/10072 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8463 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/8414 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/10686 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8655 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/10072 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/8551 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/10686 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/8915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/10227 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🧪 webkitpy~~](https://ews-build.webkit.org/#/builders/6/builds/8538 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/10686 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/8915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/10227 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/10686 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/8915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/10227 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8306 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7595 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/8915 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2032 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11805 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/8051 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsTypesGL.h:
```diff
@@ -60,6 +60,7 @@ typedef GCGLuint PlatformGLObject;
 using GCGLDisplay = void*;
 using GCGLConfig = void*;
 using GCGLContext = void*;
+using GCGLImage = void*;
```
Not here, but down there where the GCGLSync is defined.

I.e. GCGLDisplay is nothing we transfer.


GCGLSync and GCGLImage are something we transfer.

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -42,14 +42,18 @@
 #endif
 
-
+enum class GCGLExternalImage : uint8_t;
```
Problems:
- The name `GCGLExternalImage` would give impression it's the object, when it's an enumeration for the type
- ifdefing functions, manual
- move away from using GCGLints and such when normal WebCore types exists (IntSize)
- That JSIPCBinding seems like something is fishy. We are already transfering pointers as values, in GCGLsync.
- ExternalImageAttachment should not be optional, rather the return value should be optional<..>.
 
Instead, :
- other platforms can easily opt in once they need
- generator generates the code
- use webcore types where applicable
- etc

```
#ifdef PLATFORM(COCOA)
struct ExternalImageSourceSharedTextureIOSurfaceHandle  {
    MachSendRight handle;
};
struct ExternalImageSourceSharedTextureHandle {
   MachSendRight handle;
};
using ExternalImageSource = std::variant<ExternalImageSourceSharedTextureIOSurfaceHandle, ExternalImageSourceSharedTextureHandle >;
#else
using ExternalImageSource = int;
#endif

...
   using ExternalImageAttachResult = std::tuple<GCGLImage, IntSize>;
   virtual std::optional<ExternalImageAttachmentResult> attachExternalImage(GCGLenum, ExternalImageSource) = 0;
   virtual destroyExternalImage(GCGLImage) = 0;
    // ========== Other functions.
    
#if PLATFORM(COCOA)
   // Short-term support for in-process WebGL.
   virtual std::optional<ExternalImageAttachmentResult> attachExternalImage(GCGLenum, IOSurface&) = 0;
#endif
```

Alternatively call the function "create and bind external image".

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsTypesGL.h:
```diff
@@ -76,6 +77,11 @@ enum class GCGLErrorCode : uint8_t {
 };
 using GCGLErrorCodeSet = OptionSet<GCGLErrorCode>;
 
+enum class GCGLExternalImage : uint8_t {
```
This should be taken care of with the std::variant

kkinnunen-apple commented on Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
```diff
@@ -112,6 +112,12 @@ bool GraphicsContextGLANGLE::platformInitialize()
 
+void GraphicsContextGLANGLE::detachExternalImage(GCGLImage handle)
+{
+    auto display = platformDisplay();
```
I'd prefer

```
EGL_DestroyImageKHR(platformDisplay(), handle);
```

On the count of, you wouldn't use this either:
```
auto display = m_display;
EGL_DestroyImageKHR(display, handle);
```

djg replied:
I wrote `EGL_DestroyImageKHR(platformDisplay(), handle);` (which was different to the original code that I renamed) and thought `nah, that won't be accepted by Kimmo`. I was wrong. :-D

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -665,27 +665,25 @@ static bool needsEAGLOnMac()
 
 #if !PLATFORM(IOS_FAMILY_SIMULATOR)
-GraphicsContextGLCocoa::IOSurfaceTextureAttachment GraphicsContextGLCocoa::attachIOSurfaceToSharedTexture(GCGLenum target, IOSurface* surface)
+static GraphicsContextGL::ExternalImageAttachment attachMTLTextureToExternalImage(GCGLDisplay platformDisplay, GCGLenum target, MTLSharedTextureHandle* handle)
```
This could be still called something about "shared texture" instead of "mtltexture", as it's using the shared texture instead of normal mtl texture.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -708,13 +706,58 @@ static bool needsEAGLOnMac()
+    case GCGLExternalImage::IOSurface: {
+        auto surface = IOSurface::createFromSendRight(WTFMove(imageSendRight));
+        if (!surface) {
+            LOG(WebGL, "Unable to create an IOSurface from imagePort in attachExternalImage.");
```
maybe we could remove the excessive logging. The log messages are already not consistent with the code.

djg replied:
I was copying the existing style; when in doubt make it look like what was there.

I agree that the logging is inconsistent with the rest of the code.

kkinnunen-apple replied:
With inconsistency, I meant that it refers to, for example, "imagePort" which doesn't exist. Feel free to choose, match the existing logging or not.

kkinnunen-apple commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -254,7 +254,7 @@ void WebXROpaqueFramebuffer::endFrame()
         if (m_ioSurfaceTextureHandleIsShared) {
 #if !PLATFORM(IOS_FAMILY_SIMULATOR)
-            gCGL->detachIOSurfaceFromSharedTexture(m_ioSurfaceTextureHandle);
+            gCGL->detachExternalImage(m_ioSurfaceTextureHandle);
```
I believe if you're now abstracting the call, you want to invoke it through gl and not through gCGL.

kkinnunen-apple commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -137,7 +137,7 @@ void WebXROpaqueFramebuffer::startFrame(const PlatformXR::Device::FrameData::Lay
     if (data.isShared) {
 #if !PLATFORM(IOS_FAMILY_SIMULATOR)
-        auto surfaceTextureAttachment = gCGL->attachIOSurfaceToSharedTexture(textureTarget, data.surface.get());
+        auto surfaceTextureAttachment = gCGL->attachIOSurfaceToExternalImage(textureTarget, data.surface.get());
```
I believe if you're now abstracting the call, you want to invoke it through gl and not through gCGL.

So you would have:
```
gl.attachToExternalImage(textureTarget,  *data.surface);
```

The in-process variant of `GraphicsContextGL::attachToExternalImage(..., WebCore::IOSurface&)` needs to be implemented twice: 
- once for current code-path in GraphicsContextGLCocoa
- once for RemoteGraphicsContextGLProxyCocoa, obtaining the mach send right from the iosurface.

djg replied:
But we can't create an in process IOSurface when RemoteGraphicsContextGLProxyCocoa is in use, can we?

kkinnunen-apple replied:
> But we can't create an in process IOSurface when RemoteGraphicsContextGLProxyCocoa is in use, can we?

Yeah, we can. So "WebGL GPUP Yes/No" is distinct choice to "IOKit blocking".

Note: the question is not logical, there is need in *creating* in-process IOSurface for RemoteGraphicsContextGLProxyCocoa.

The client gives use either send right to IOSurface or the IOSurface instance.

The cases with RemoteGraphicsContextGLProxyCocoa are:
 - If we already have in process IOSurface, we can give that to RemoteGraphicsContextGLProxyCocoa via obtaining send right to it.
 - If we already have Mach send right to a IOSurface, we can give that to RemoteGraphicsContextGLProxyCocoa.

kkinnunen-apple commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
```diff
@@ -126,50 +124,49 @@ void WebXROpaqueFramebuffer::startFrame(const PlatformXR::Device::FrameData::Lay
     if (data.isShared) {
 #if !PLATFORM(IOS_FAMILY_SIMULATOR)
-        auto surfaceTextureAttachment = gCGL->attachIOSurfaceToSharedTexture(textureTarget, data.surface.get());
+        auto surfaceTextureAttachment = gCGL->createAndBindExternalImage(textureTarget, data.surface.get());
```
here we might still use the abstracted interface via `gl.` unless osmething else prevents that.

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsTypesGL.h:
```diff
@@ -90,4 +96,12 @@ template <> struct EnumTraits<GCGLErrorCode> {
     >;
 };
 
+template <> struct EnumTraits<GCGLExternalImage> {
```
Maybe you forgot the remove  GCGLExternalImage

kkinnunen-apple commented on Tools/Scripts/generate-gpup-webgl:
```diff
@@ -89,6 +89,9 @@ context_messages_template = (
 
 messages -> RemoteGraphicsContextGL NotRefCounted Stream {{
+#if PLATFORM(COCOA)
+    void CreateAndBindExternalImage(GCGLenum target, WebCore::ExternalImageSource source) -> (uint64_t handle, WebCore::IntSize size) Synchronous
```
possibly this will be changed by the learning experience, e.g this can go to the autogenerated section.

djg replied:
This can't be generated automatically because `ExternalImageSource` contains `WTF::MachSendRight` and serializers can't be generated for `StreamConnectionEncoder`.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -680,56 +680,86 @@ static bool needsEAGLOnMac()
+    MTLSharedTextureHandle* handle = nil;
+    return WTF::switchOn(source,
+        [&](const ExternalImageSourceIOSurfaceHandle& ioSurface) -> std::optional<ExternalImageAttachResult> {
+            auto surface = IOSurface::createFromSendRight(WTFMove(ioSurface.handle));
```
this looks somehow odd, e.g. you move internals of the const ref object into the IOSurface::createFromSendRight.
This works because of bug in  IOSurface::createFromSendRight type signature.

So you should have form of:
```
return WTF::switchOn(WTFMove(source),
         [&](ExternalImageSourceIOSurfaceHandle&& ioSurface) -> std::optional<ExternalImageAttachResult> {
          ....
        }, 
                 [&](ExternalImageSourceMTLSharedTextureHandle&& sharedTexture) -> std::optional<ExternalImageAttachResult> {
                  ....
       }
```

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsContextGL.h:
```diff
@@ -1681,6 +1708,38 @@ IMPLEMENT_GCGL_OWNED(Texture)
 
 #undef IMPLEMENT_GCGL_OWNED
 
+#if PLATFORM(COCOA)
```
If you want to style this, I think this can be declared in the Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in

```
#if ENABLE(GPU_PROCESS) && ENABLE(WEBGL)
....
header: <WebCore/GraphicsContextGL.h>
#if PLATFORM(COCOA)
struct WebCore::ExternalImageSourceIOSurfaceHandle {
      MachSendRight handle;
};
struct WebCore::ExternalImageSourceMTLSharedTextureHandle {
     MachSendRight handle;
};
#endif

```

djg replied:
Excellent.

kkinnunen-apple commented on Source/WebCore/platform/graphics/GraphicsTypesGL.h:
```diff
@@ -47,6 +47,7 @@ typedef float GCGLclampf;
 typedef char GCGLchar;
 typedef void* GCGLsync;
 typedef void GCGLvoid;
+typedef void* GCGLimage;
```
This might be part of the learning thing, you might mark this up as GCEGLimage.

djg replied:
Yes.

djg committed:
[WebGL] Abstract attaching/detaching external images
https://bugs.webkit.org/show_bug.cgi?id=256945
rdar://109495127

Reviewed by Kimmo Kinnunen.

When the GPU process is enabled, the Web process no long has access to the APIs
required to create images passed from the UI process. This change added APIs to
allow attaching and detaching external images that can be represented by a
MachSendRight (either IOSurface or MTLSharedTexture), that can be used in the
Web or GPU process.

Support work for WebXR in GPUP.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
* Source/WebCore/PAL/pal/spi/cocoa/MetalSPI.h:
* Source/WebCore/platform/graphics/GraphicsContextGL.h:
* Source/WebCore/platform/graphics/GraphicsTypesGL.h:
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.cpp:
(WebCore::GraphicsContextGLANGLE::createAndBindExternalImage):
(WebCore::GraphicsContextGLANGLE::destroyEGLImage):
* Source/WebCore/platform/graphics/angle/GraphicsContextGLANGLE.h:
* Source/WebCore/platform/graphics/cocoa/ANGLEUtilitiesCocoa.mm:
(WebCore::newSharedEventWithMachPort):
(WebCore::newSharedEvent):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::createAndBindSharedTextureToExternalImage):
(WebCore::GraphicsContextGLCocoa::createAndBindExternalImage):
(WebCore::GraphicsContextGLCocoa::attachIOSurfaceToSharedTexture): Deleted.
(WebCore::GraphicsContextGLCocoa::detachIOSurfaceFromSharedTexture): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.cpp:
(WebKit::RemoteGraphicsContextGL::createAndBindExternalImage):
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.h:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGL.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteGraphicsContextGLFunctionsGenerated.h:
(destroyEGLImage):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.mm:
(IPC::ArgumentCoder<WebCore::GraphicsContextGL::ExternalImageSourceIOSurfaceHandle>::encode):
(IPC::ArgumentCoder<WebCore::GraphicsContextGL::ExternalImageSourceIOSurfaceHandle>::decode):
(IPC::ArgumentCoder<WebCore::GraphicsContextGL::ExternalImageSourceMTLSharedTextureHandle>::encode):
(IPC::ArgumentCoder<WebCore::GraphicsContextGL::ExternalImageSourceMTLSharedTextureHandle>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.cpp:
(WebKit::RemoteGraphicsContextGLProxy::createAndBindExternalImage):
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteGraphicsContextGLProxyFunctionsGenerated.cpp:
(WebKit::RemoteGraphicsContextGLProxy::destroyEGLImage):
* Tools/Scripts/generate-gpup-webgl:

Canonical link: https://commits.webkit.org/264712@main
